pymc/tests/test_model.py

```diff
@@ -702,6 +703,61 @@ def test_nested_model_coords():
     assert set(m2.RV_dims) < set(m1.RV_dims)
 
 
+def test_shape_error_from_fixed_dims():
+    """
+    with #5751 dimensions set by add_coord() are default TensorConstant type
+    and so have fixed dimensions. Trying to resize them will raise a ShapeError,
+    which this test checks for.
+    """
+    with pm.Model() as pmodel:
+        a = pm.Normal("a", mu=[1, 2, 3], dims="fixed")
+        m = pm.MutableData("m", [[1, 2, 3]], dims=("one", "fixed"))
+
+    with pytest.raises(ShapeError):
+        # Can't work because the "fixed" dimension is linked to a constant shape:
+        # Note that the new data tries to change both dimensions
+        with pmodel:
+            pm.set_data({"m": [[1, 2], [3, 4]]})
+
+
+def test_set_data_with_fixed_dims():
+    """
+    test that set_data doesn't raise ShapeError with fixed dims when
+    correct dimensions passed.
+    """
+    with pm.Model() as pmodel:
+        a = pm.Normal("a", mu=[1, 2, 3], dims="fixed")
+        m = pm.MutableData("m", [[1, 2, 3]], dims=("one", "fixed"))
+
+        # This is fine because the "fixed" dim is not resized
+        pm.set_data({"m": [[1, 2, 3], [3, 4, 5]]})
+
+
+def test_add_coord_fixed_dims():
+    """
+    test that dims does not cause resizing of y. See #5181.
+    """
+    with pm.Model(coords={"feature": [1], "group": ["A", "B"]}):
+        x = pm.Normal("x", 0, 1, dims="feature")
+        y = pm.Normal("y", x, 1, dims=("group", "feature"))
+    assert x.eval().shape == (1,)
+    assert y.eval().shape == (2, 1)
```

Comment:
This is the test case that can be used to assert the `ShapeWarning`:

```python
with pytest.warns(ShapeWarning, match="was initialized from coords"):
    ...
```

Reply:
This code doesn't raise a ShapeWarning. Should it?

Reply:
Ah no worries, figured it out.

Reply:
@michaelosthege I'm struggling to isolate the ShapeWarning, the code 
```python
with pm.Model() as pmodel:
    pmodel.add_coord("shared", [1,2,3], fixed=False)
    m = pm.MutableData("m", [1,2,3], dims=("shared"))

    pm.set_data({"m": [1,2,3,4]}) 
```
raises both ShapeWarning and a ValueError
```
ShapeWarning: You're changing the shape of a variable in the 'shared' dimension which was 
initialized from coords. Make sure to update the corresponding coords, otherwise you'll get shape issues.
  warnings.warn(

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/yw/hvmlf6sd7yv3040wn8str6540000gn/T/ipykernel_1859/831241587.py in <module>
      4     m = pm.MutableData("m", [1,2,3], dims=("shared"))
      5 
----> 6     pm.set_data({"m": [1,2,3,4]})

pymc/pymc/model.py in set_data(new_data, model, coords)
   1798 
   1799     for variable_name, new_value in new_data.items():
-> 1800         model.set_data(variable_name, new_value, coords=coords)
   1801 
   1802 

pymc/pymc/model.py in set_data(self, name, values, coords)
   1220                 if original_coords is not None:
   1221                     if new_coords is None:
-> 1222                         raise ValueError(
   1223                             f"The '{name}' variable already had {len(original_coords)} coord values defined for "
   1224                             f"its {dname} dimension. With the new values this dimension changes to length "

ValueError: The 'm' variable already had 3 coord values defined for its shared dimension. 
With the new values this dimension changes to length 4, so new coord values for the shared dimension are required.
```
I also tried the example from #5760,
```python
x = np.random.normal([13,14,12],[2,1,3],size=(100,3))
x_new = x = np.random.normal([14,15,11],[2,1,3],size=(100,3))

with pm.Model(coords=coords) as model:
    model.add_coord("category", ["A","B","C"], fixed = False)
    model.add_coord("data_point", list(range(1,101)), fixed = False)
    x = pm.MutableData("x",x,dims=["data_point","country"])
    mu = pm.Normal("mu",mu=[12,12,12],sigma=2)
    sigma = pm.Exponential("sigma",lam=[1,1,1])
    obs = pm.Normal("obs",mu=mu,sigma=sigma,observed=x)
with model:
    pm.set_data({"x":x_new})
```
Which doesn't raise any errors or warnings. Not sure how to proceed here...

Reply:
The second example doesn't resize the `"data_point"` dimension, that's why it works without problems.

The traceback you posted above looks like `set_data` first warns, but then proceeds to raise a value error anyhow.

Obviously I overlooked this code path in #5761.
We should try to answer the following question:

__Is there a valid scenario where someone might want to resize a dimension that was initially created from `coords`, but can not provide new coordinate values in the same call to `set_data`?__

I feel like with the `ShapeWarning` we assume that there is such a scenario, so we should come up with a corresponding test.
But if in fact this scenario does not exist, we can just raise the `ValueError` directly.

Reply:
@michaelosthege I would say we should raise the ValueError directly for this scenario. I can't imagine a situation wherein someone would both be using `add_coord()` directly and then not be able to use the coord they had previously passed to `add_coord()` in `set_data()`.

Reply:
Agreed. And let's make sure to have `match="..."` for all the error/warning asserts :)